v2/snmp/manager.go

```diff
@@ -29,32 +35,181 @@ type Response struct {
 // If the function returns an error, the walk will be terminated.
 type Walker func(pdu *PDU) error
 
-// PDU defines ...
+// PDU defines an SNMP PDU.
 type PDU struct {
-	// TBD
+	RequestID   int32
+	Error       int
+	ErrorIndex  int
+	VarbindList []Varbind
 }
 
 type managerImpl struct {
-	conn   net.Conn
-	config *managerConfig
+	conn          net.Conn
+	config        *managerConfig
+	nextRequestID int32
 }
 
-func (m *managerImpl) Get(ctx context.Context, oids []string) (*Response, error) {
-	// TODO
-	return nil, nil
+type Varbind struct {
+	OID   asn1.ObjectIdentifier
+	Value interface{}
+}
+type packet struct {
+	Version     SNMPVersion
+	Community   []byte
+	RequestType asn1.RawValue
 }
 
-func (m *managerImpl) GetNext(ctx context.Context, oids []string) (*Response, error) {
-	// TODO
-	return nil, nil
+const maxInputBufferSize = 65535
+
+type messageType byte
+
+const getMessage = 0xA0
+const getNextMessage = 0xA1
+const getBulkMessage = 0xA5
+
+func (m *managerImpl) Get(ctx context.Context, oids []string) (*PDU, error) {
+	return m.executeGet(ctx, getMessage, 0, 0, oids)
 }
 
-func (m *managerImpl) GetBulk(ctx context.Context, oids []string, nonRepeaters int, maxRepetitions int) (*Response, error) {
-	// TODO
-	return nil, nil
+func (m *managerImpl) GetNext(ctx context.Context, oids []string) (*PDU, error) {
+	return m.executeGet(ctx, getNextMessage, 0, 0, oids)
+}
+
+func (m *managerImpl) GetBulk(ctx context.Context, oids []string, nonRepeaters, maxRepetitions int) (*PDU, error) {
+	return m.executeGet(ctx, getBulkMessage, nonRepeaters, maxRepetitions, oids)
 }
 
 func (m *managerImpl) GetWalk(ctx context.Context, oid string, walker Walker) error {
 	// TODO
 	return nil
 }
+
+func (m *managerImpl) executeGet(ctx context.Context, mType messageType, nonRepeaters, maxRepetitions int, oids []string) (*PDU, error) {
+	ctx, cancel := context.WithTimeout(ctx, m.config.timeout)
+	defer cancel()
+	deadline, _ := ctx.Deadline()
+	err := m.conn.SetDeadline(deadline)
+	if err != nil {
+		return nil, err
+	}
+
+	b, err := m.buildPacket(oids, mType, nonRepeaters, maxRepetitions)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.writePacket(b)
+	if err != nil {
+		return nil, err
+	}
+
+	input, err := m.readResponse()
+	if err != nil {
+		// TODO Handle EOF
+		return nil, err
+	}
+
+	return m.parseResponse(input)
+}
+
+func (m *managerImpl) writePacket(b []byte) (err error) {
+	n, err := m.conn.Write(b)
+	m.config.trace.WriteComplete(m.config, b[0:n], err)
+	return
+}
+
+func (m *managerImpl) readResponse() (input []byte, err error) {
+	input = make([]byte, maxInputBufferSize)
+
+	n, err := m.conn.Read(input[:])
+	defer m.config.trace.ReadComplete(m.config, input[0:n], err)
+	if err != nil {
+		return nil, err
+	}
+
+	if n == maxInputBufferSize {
+		// Never expect this to happen
+		panic(fmt.Errorf("overflowing response buffer"))
```

Comment:
accepting that this should never happen, is panic the right response rather than a user-generated error